### ISSUE TITLE
[React] Add "capture" option to useDismiss

### ISSUE THREAD

**Issue by YanZisuka** [floating-ui/floating-ui#2594]
**Is your feature request related to a problem? Please describe.**
When a modal dialog appears, we stop propagating events in the modal overlay because we don't want it to interact with the UI in the background.

In this case, if the modal's child element has a component that uses useDismiss, the dismiss event will be interrupted before it reaches the document and `closeOnPressOutside` will not work, so I want to use capture to detect the click.

**Describe the solution you'd like**
We can check the implementation at [airbnb/react-outside-click-handler
](https://github.com/airbnb/react-outside-click-handler/blob/master/src/OutsideClickHandler.jsx#L29)

so, in [useDismiss.ts](https://github.com/floating-ui/floating-ui/blob/master/packages/react/src/hooks/useDismiss.ts).
Briefly, consider the following code:

```tsx
export interface UseDismissProps {
  enabled?: boolean;
  escapeKey?: boolean;
  ...
  capture?: boolean | {escapeKey?: boolean; outsidePress?: boolean}; // new line
}

...

const doc = getDocument(floating);
    escapeKey && doc.addEventListener('keydown', closeOnEscapeKeyDown, {capture: escapeKeyCapture});
    outsidePress &&
      doc.addEventListener(outsidePressEvent, closeOnPressOutside, {capture: outsidePressCapture});
```

Thank you for the nice library and dedication. :)

**Describe alternatives you've considered**
N/A

**Additional context**
[Capture phase events in React's official docs](https://react.dev/learn/responding-to-events#capture-phase-events)

**Comment by atomiks** [comment]
There's an open PR already: https://github.com/floating-ui/floating-ui/pull/2345

But it has a problem, because capture phase is used to detect if "outside press" originated inside the React tree (i.e., in a nested portaled floating element) even if it's outside of the DOM tree.

**Comment by YanZisuka** [comment]
Thanks for your reply!

Well, I don't think I fully understand the problem yet.

But I think the behavior you're talking about is consistent with the design intent, since React doesn't want programmers to think much about the actual DOM.

**Comment by atomiks** [comment]
React's event system bubbles through portaled React elements, which allows us to determine if a nested floating element (by the React tree) has been pressed, even when it's not a child of the parent in the DOM. This prevents it from closing unexpectedly.

The technique used to do this uses capture phase so it runs before the bubble phase. If we switch the main handler to capture too, then this fires afterwards and breaks the logic.

https://github.com/floating-ui/floating-ui/blob/851e8a0d56041cb6100388a065b3a0edc91dc5ac/packages/react/src/hooks/useDismiss.ts#L329-L333

So a different/better technique is needed for this not to occur.

**Comment by YanZisuka** [comment]
I forked this project and [implemented a simple concept](https://github.com/YanZisuka/floating-ui/pull/1/files#diff-fc836a8746308da2586828c2a4cbfbf8accc236b95489446f3cd4c4b634eb319R285-R291).

In the document capture phase, a main handler is added to `event.target`, and during the "AT_TARGET" phase, the handler is executed to operate the logic that determines `insideReactTree`.

I passed the test, what do you think about this?

**Comment by atomiks** [comment]
@YanZisuka interesting solution! If the tests pass then it's all good, so feel free to make a PR. I'll test it out a bit further as well.

**Comment by atomiks** [comment]
Closed by #2597
